detox/src/devices/drivers/android/genycloud/GenyCloudDriver.js

```diff
@@ -0,0 +1,66 @@
+const _ = require('lodash');
+const AndroidDriver = require('../AndroidDriver');
+const GenyCloudExec = require('./exec/GenyCloudExec');
+const RecipesService = require('./services/GenyRecipesService');
+const InstanceLookupService = require('./services/GenyInstanceLookupService');
+const InstanceLifecycleService = require('./services/GenyInstanceLifecycleService');
+const InstanceNaming = require('./services/GenyInstanceNaming');
+const AllocationHelper = require('./helpers/GenyAllocationHelper');
+const DeviceQueryHelper = require('./helpers/GenyDeviceQueryHelper');
+const DetoxRuntimeError = require('../../../../errors/DetoxRuntimeError');
+const logger = require('../../../../utils/logger').child({ __filename });
+
+class GenyCloudDriver extends AndroidDriver {
+  constructor(config) {
+    super(config);
+    this._name = 'Unspecified Genymotion Cloud Emulator';
+
+    const exec = new GenyCloudExec();
+    const instanceNaming = new InstanceNaming(); // TODO should consider a permissive impl for debug/dev mode. Maybe even a custom arg in package.json (Detox > ... > genycloud > sharedAccount: false)
+    this.recipeService = new RecipesService(exec, logger);
+    this.instanceLookupService = new InstanceLookupService(exec, instanceNaming, this.deviceRegistry);
+    this.instanceLifecycleService = new InstanceLifecycleService(exec, instanceNaming);
+    this.deviceQueryHelper = new DeviceQueryHelper(this.recipeService);
+    this.allocationHelper = new AllocationHelper(this.instanceLookupService, this.instanceLifecycleService);
+  }
+
+  get name() {
+    return this._name;
+  }
+
+  async acquireFreeDevice(deviceQuery) {
+    const recipe = await this.deviceQueryHelper.getRecipeFromQuery(deviceQuery);
+    this.assertRecipe(deviceQuery, recipe);
+
+    const cookie = { coldBoot: false };
+    const adbName = await this.allocateDevice({
+      recipe,
+      cookie,
+      toString: () => recipe.toString(),
+    });
+
+    await this.emitter.emit('bootDevice', { coldBoot: cookie.coldBoot, deviceId: adbName, type: recipe.name});
+    await this.adb.apiLevel(adbName);
+    await this.adb.disableAndroidAnimations(adbName);
+
+    this._name = `GenyCloud-${adbName}`;
+    return adbName;
+  }
+
+  async doAllocateDevice({ recipe, cookie }) {
+    const { instance, coldBooted } = await this.allocationHelper.allocateInstance(recipe.uuid);
+    cookie.coldBoot = coldBooted;
+    return instance.adbName;
+  }
+
+  assertRecipe(deviceQuery, recipe) {
+    if (!recipe) {
+      throw new DetoxRuntimeError({
+        message: 'No Genycloud devices found for recipe!',
```

Comment:
The error does not look really helpful.
I'd dump the `recipes list` output here to `debugInfo` variable.
It is a bit long, sure, that's about 20KB, but that would really help users since anyway the tests won't go further so why not?

```
UUID                                  NAME                     ANDROID    SCREEN               SOURCE
------------------------------------  -----------------------  ---------  -------------------  ----------
b1e51193-4db8-43f7-b7cd-ae35eebb6bca  Custom Phone             4.4.4      768 x 1280 dpi 320   genymotion
5e4c60b3-327b-4fb4-98e2-4f996900e145  Custom Tablet            4.4.4      1536 x 2048 dpi 320  genymotion
42b60b69-bd60-4c3d-912c-4073d854573d  Google Nexus 10          4.4.4      2560 x 1600 dpi 320  genymotion
31f96174-9148-4bd3-84dd-0819681936bf  Google Nexus 4           4.4.4      768 x 1280 dpi 320   genymotion
b4f164b3-277e-4efe-9fcd-38f6d847ddd7  Google Nexus 5           4.4.4      1080 x 1920 dpi 480  genymotion
e38a1ec9-5fbf-481c-9103-d1782ea9f1d7  Google Nexus 7           4.4.4      800 x 1280 dpi 213   genymotion
3ca42f69-ff08-42b3-8d78-2ccb617e55f2  Google Nexus 7 2013      4.4.4      1200 x 1920 dpi 320  genymotion
2f71872c-8bbc-4260-a284-7c6f50ede169  HTC One                  4.4.4      1080 x 1920 dpi 480  genymotion
47f68b9a-08cb-4b55-b0a7-a6664d118829  Motorola Moto X          4.4.4      720 x 1280 dpi 320   genymotion
de20111c-332a-4cb4-8088-d4a7f8f961ec  Samsung Galaxy Note 2    4.4.4      720 x 1280 dpi 320   genymotion
c1fb5618-68de-47c6-9659-e28a30de156c  Samsung Galaxy Note 3    4.4.4      1080 x 1920 dpi 480  genymotion
1df05208-b43d-4ac4-a0af-1b60d7efd763  Samsung Galaxy S3        4.4.4      720 x 1280 dpi 320   genymotion
15edb3ca-3bf4-4778-a85c-c04a7d005715  Samsung Galaxy S4        4.4.4      1080 x 1920 dpi 480  genymotion
4a7cf261-fa75-4a18-a564-718ebefba390  Samsung Galaxy S5        4.4.4      1080 x 1920 dpi 480  genymotion
3302ec22-38d0-4ccb-9f6d-6bea2e1ede13  Sony Xperia Tablet Z     4.4.4      1920 x 1200 dpi 320  genymotion
ca245ccd-6920-4b3b-a354-2592887a7bb4  Custom Phone             5.0.0      768 x 1280 dpi 320   genymotion
f6db2e31-feee-4262-ae4b-dd3749eece87  Custom Tablet            5.0.0      1536 x 2048 dpi 320  genymotion
f1a8a52b-5b1c-4d24-806f-3fbe838eb32d  Google Nexus 10          5.0.0      2560 x 1600 dpi 320  genymotion
d39db595-6ae3-4ae9-bddc-fcaf85cf1d83  Google Nexus 4           5.0.0      768 x 1280 dpi 320   genymotion
9c667b7a-1a43-4d4f-8b5b-2dc8c7888d21  Google Nexus 5           5.0.0      1080 x 1920 dpi 480  genymotion
b7347f03-2a75-45a1-a8a1-42324198125e  Google Nexus 6           5.0.0      1440 x 2560 dpi 560  genymotion
57110e09-24f5-4e32-b4ea-a549e95970bf  Google Nexus 7           5.0.0      800 x 1280 dpi 213   genymotion
8e33f3db-68d9-4700-95dd-850e23d634ed  Google Nexus 7 2013      5.0.0      1200 x 1920 dpi 320  genymotion
4f50f627-a4c8-4368-bc5c-91467a9ceb24  Google Nexus 9           5.0.0      1536 x 2048 dpi 320  genymotion
7fad4b10-140c-4227-84d0-22dd88c8f8b1  Samsung Galaxy S6        5.0.0      1440 x 2560 dpi 640  genymotion
84f0cceb-c1a2-4380-a1a7-1e4ae3f74a4b  Sony Xperia Tablet Z     5.0.0      1920 x 1200 dpi 320  genymotion
3752d651-4657-4284-b620-978472e7d8ac  Custom Phone             5.1.0      768 x 1280 dpi 320   genymotion
cde39d84-1698-47f8-8edb-b0c1dda09bcb  Custom Tablet            5.1.0      1536 x 2048 dpi 320  genymotion
c6e1222c-993e-4bf2-840f-4795792f2143  Google Nexus 10          5.1.0      2560 x 1600 dpi 320  genymotion
55fddf0b-0e56-4742-8791-106ad93c01ff  Google Nexus 4           5.1.0      768 x 1280 dpi 320   genymotion
e70d130c-955c-4f25-a451-0ed2e57a24ac  Google Nexus 5           5.1.0      1080 x 1920 dpi 480  genymotion
a9af19b8-8df9-41a8-81ac-a6294648ecc1  Google Nexus 6           5.1.0      1440 x 2560 dpi 560  genymotion
89be4b14-a11a-471c-a3a5-2a9237a80007  Google Nexus 7           5.1.0      800 x 1280 dpi 213   genymotion
8616f424-caac-4dc9-864d-d1381b1b6360  Google Nexus 7 2013      5.1.0      1200 x 1920 dpi 320  genymotion
8bdeb4ca-7114-4920-903d-b33085b38eaa  Google Nexus 9           5.1.0      1536 x 2048 dpi 320  genymotion
865d12b1-d839-49a7-90cf-175cab4ba51f  Samsung Galaxy S6        5.1.0      1440 x 2560 dpi 640  genymotion
76e0cd80-7904-4d41-a812-47f9598ded23  Sony Xperia Tablet Z     5.1.0      1920 x 1200 dpi 320  genymotion
8b8e7929-cb0e-4aac-b367-aa91ce5b506d  Amazon Fire 7            6.0.0      1024 x 600 dpi 160   genymotion
f028a901-70e3-4d31-8aca-3993d2dbcde3  Custom Phone             6.0.0      768 x 1280 dpi 320   genymotion
97027cf3-9498-4bf1-8414-88a6b3d4b83c  Custom Tablet            6.0.0      1536 x 2048 dpi 320  genymotion
9fc4abf6-4a22-443d-b045-d3ef2bd96974  Google Nexus 5           6.0.0      1080 x 1920 dpi 480  genymotion
c660c8c2-df75-44ac-9a81-e08f468f1c10  Google Nexus 5X          6.0.0      1080 x 1920 dpi 420  genymotion
da95cf5c-7652-46ce-a946-906d93fc2eca  Google Nexus 6           6.0.0      1440 x 2560 dpi 560  genymotion
46028057-70cd-49f5-9c91-d7413b4e4400  Google Nexus 6P          6.0.0      1440 x 2560 dpi 560  genymotion
1f173992-993a-4241-9db8-f68685c7f404  Google Nexus 7 2013      6.0.0      1200 x 1920 dpi 320  genymotion
445e774c-a9e9-49bd-84b9-673c4e9f7163  Google Nexus 9           6.0.0      1536 x 2048 dpi 320  genymotion
35d91457-dbfe-47ac-98fc-3271440753c4  Google Pixel C           6.0.0      2560 x 1800 dpi 320  genymotion
082aa083-451d-4d4d-a680-b784d2012b94  Samsung Galaxy S6        6.0.0      1440 x 2560 dpi 640  genymotion
a2a0c86c-7572-45fe-98d0-66f8efed9fa0  Samsung Galaxy S7        6.0.0      1440 x 2560 dpi 560  genymotion
23cf0c50-e80b-4fc8-ab9c-fa767b806ee9  Custom Phone             7.0.0      768 x 1280 dpi 320   genymotion
bb0a0ca4-ddd8-449f-9f16-093f7351b9cf  Custom Tablet            7.0.0      1536 x 2048 dpi 320  genymotion
43ea364a-efb0-41c2-99b9-b3f269fea1cc  Google Nexus 5X          7.0.0      1080 x 1920 dpi 420  genymotion
a0a9c90a-b391-42f4-b77b-ae0561d74bbe  Google Nexus 6           7.0.0      1440 x 2560 dpi 560  genymotion
9f1adea8-e280-460d-9319-580570f61e8c  Google Nexus 6P          7.0.0      1440 x 2560 dpi 560  genymotion
f5ba679d-bc43-4b33-b046-4923835851b1  Google Nexus 9           7.0.0      1536 x 2048 dpi 320  genymotion
b747b286-4729-46ef-9bfa-f05942b15588  Google Pixel C           7.0.0      2560 x 1800 dpi 320  genymotion
8f373fa1-97fa-4652-84c8-a006459d50a0  Samsung Galaxy S8        7.0.0      1440 x 2960 dpi 480  genymotion
5f6d4c94-d926-42cc-8019-f902295fc36d  Amazon Fire 7            7.1.0      1024 x 600 dpi 160   genymotion
45f37afb-0a08-4775-8358-09d7ccc15adb  Amazon Fire HD 8         7.1.0      1280 x 800 dpi 213   genymotion
c41a277f-39a1-402c-bf92-34d86732f89d  Custom Phone             7.1.0      768 x 1280 dpi 320   genymotion
5fdbb58a-b612-4cdd-939b-15e9982b69c9  Custom Tablet            7.1.0      1536 x 2048 dpi 320  genymotion
4b13441b-dacc-4803-b2e3-cb2d8d23def2  Google Nexus 5X          7.1.0      1080 x 1920 dpi 420  genymotion
80a67ae9-430c-4824-a386-befbb19518b9  Google Nexus 6           7.1.0      1440 x 2560 dpi 560  genymotion
00cd578a-14de-4f47-b4f1-d454d613d9da  Google Nexus 6P          7.1.0      1440 x 2560 dpi 560  genymotion
bccf86e5-7280-47bf-a502-6f64eba116f7  Google Nexus 9           7.1.0      1536 x 2048 dpi 320  genymotion
de849ff1-e051-4844-a92f-51b0181a6586  Google Pixel             7.1.0      1080 x 1920 dpi 420  genymotion
bb06e95f-b326-4510-a8e5-22c0f9ed9e83  Google Pixel C           7.1.0      2560 x 1800 dpi 320  genymotion
48007128-c74d-461a-892a-f88aa3c1a694  Google Pixel XL          7.1.0      1440 x 2560 dpi 560  genymotion
6185bb7c-1f96-40bb-9f6c-798bd7547d36  Samsung Galaxy S6        7.1.0      1440 x 2560 dpi 640  genymotion
d777f454-c238-4a59-a587-6b88efcd5986  Samsung Galaxy S7        7.1.0      1440 x 2560 dpi 560  genymotion
2b104999-1937-401e-b9b8-7a9a857d5088  Custom Phone             8.0        768 x 1280 dpi 320   genymotion
dd55a4aa-aefd-4594-b299-8d09a1392502  Custom Tablet            8.0        1536 x 2048 dpi 320  genymotion
bf1a9765-c743-4cdc-95ec-c40a74493055  Google Nexus 5X          8.0        1080 x 1920 dpi 420  genymotion
a59951f2-ed13-40f9-80b9-3ddceb3c89f5  Google Nexus 6           8.0        1440 x 2560 dpi 560  genymotion
180ddae8-436c-4ae3-a274-91144b7e4bb4  Google Nexus 6P          8.0        1440 x 2560 dpi 560  genymotion
74ad0f8b-90f5-47c5-bc7a-9c05b04de4ca  Google Pixel             8.0        1080 x 1920 dpi 420  genymotion
c52fdfc2-6914-4266-aa6e-50258f50ef91  Google Pixel 2           8.0        1080 x 1920 dpi 420  genymotion
cbd25b62-a120-4ea2-9528-36f575478775  Google Pixel 2 XL        8.0        1440 x 2880 dpi 560  genymotion
3e1aa8be-b625-4622-8705-c5a0c69fa687  Google Pixel C           8.0        2560 x 1800 dpi 320  genymotion
06867de4-4b99-4842-ba40-fd3daaabdf23  Google Pixel XL          8.0        1440 x 2560 dpi 560  genymotion
e7a4ecd9-6044-41c7-ace3-ccee5402b590  Samsung Galaxy S7        8.0        1440 x 2560 dpi 560  genymotion
b9cf7b2c-4d11-4777-97c7-29d3b5c68d59  Samsung Galaxy S8        8.0        1440 x 2960 dpi 480  genymotion
b0a2e032-2117-4066-9d74-3904eceef4a8  Samsung Galaxy S9        8.0        1440 x 2960 dpi 560  genymotion
77a64f9d-1414-4cee-9be2-2f26bbfd1dfd  Custom Phone             8.1        768 x 1280 dpi 320   genymotion
18eb59e5-accf-48d6-b62a-a8b4740f1981  Custom Tablet            8.1        1536 x 2048 dpi 320  genymotion
20ee73af-5c32-405c-8582-cd19ffd175c2  Google Pixel             8.1        1080 x 1920 dpi 420  genymotion
0203df5d-ee96-4930-be21-b55e9c445032  Google Pixel 2           8.1        1080 x 1920 dpi 420  genymotion
c5b09eca-fd2f-4eb1-a6b5-4e7d9821e3e9  Google Pixel 2 XL        8.1        1440 x 2880 dpi 560  genymotion
208e8938-2b3a-4b69-823e-ce1c5decfca9  Google Pixel C           8.1        2560 x 1800 dpi 320  genymotion
ea5fda48-fa8b-48c1-8acc-07d910856141  Google Pixel XL          8.1        1440 x 2560 dpi 560  genymotion
81122aa5-a45e-4858-9848-b23ff03a9e9d  Samsung Galaxy S8        8.1        1440 x 2960 dpi 480  genymotion
8c59e59e-283f-4dcb-bf55-6523b30b59f9  Amazon Fire HD 10        9.0        1920 x 1200 dpi 240  genymotion
df785e5b-b1bf-470b-af5d-0c28a3d98a26  Custom Phone             9.0        768 x 1280 dpi 320   genymotion
63cd126c-406e-4886-ac5d-8a7d6ca6486c  Custom Tablet            9.0        1536 x 2048 dpi 320  genymotion
107d757e-463a-4a18-8667-b8dec6e4c87e  Google Pixel             9.0        1080 x 1920 dpi 420  genymotion
e6a305b5-ca40-4587-9aa8-623eb535b2f2  Google Pixel 2           9.0        1080 x 1920 dpi 420  genymotion
dfbdd1bc-cce2-4f27-b8be-535b93ff9ee7  Google Pixel 2 XL        9.0        1440 x 2880 dpi 560  genymotion
143eb44a-1d3a-4f27-bcac-3c40124e2836  Google Pixel 3           9.0        1080 x 2160 dpi 440  genymotion
e5008049-8394-40fc-b7f8-87fa9f1c305f  Google Pixel 3 XL        9.0        1440 x 2960 dpi 560  genymotion
fb936099-d261-4dd3-8dec-3fc14f1d0f03  Google Pixel 3a          9.0        1080 x 2220 dpi 420  genymotion
bd402826-4ee6-4598-94df-da4f89021042  Google Pixel XL          9.0        1440 x 2560 dpi 560  genymotion
95c48908-744b-4768-a6c0-4fca8e27d166  Huawei P30 Pro           9.0        1080 x 2340 dpi 360  genymotion
a2631466-c3e1-4fea-b625-80ee31f384ee  Samsung A10              9.0        720 x 1520 dpi 260   genymotion
b5fb8027-05a1-4ade-937b-38b04696104e  Samsung A50              9.0        1080 x 2340 dpi 400  genymotion
a3f26fce-0632-4df4-88aa-cbddbc739e1e  Samsung Galaxy Note 10+  9.0        1440 x 3040 dpi 480  genymotion
c6c6a0f3-0b13-4411-b30b-4827fb699500  Samsung Galaxy S10       9.0        1440 x 3040 dpi 560  genymotion
f34de94f-1e85-4d61-9b0d-c47968bc156c  Samsung Galaxy S8        9.0        1440 x 2960 dpi 480  genymotion
e20da1a3-313c-434a-9d43-7268b12fee08  Samsung Galaxy S9        9.0        1440 x 2960 dpi 560  genymotion
d8b10016-c02a-41f9-8a91-ce9b44197d21  Xiaomi Redmi Note 7      9.0        1080 x 2340 dpi 420  genymotion
3a66f8a3-05ca-41e2-9f91-31c1bc311d64  Custom Phone             10.0       768 x 1280 dpi 320   genymotion
3da64198-5895-478d-9680-a51a92811e1f  Custom Tablet            10.0       1536 x 2048 dpi 320  genymotion
c259202b-6605-44eb-978c-040b2edbc364  Google Pixel 3           10.0       1080 x 2160 dpi 440  genymotion
4c015ada-e64e-4f5d-a320-06cbf6e95648  Google Pixel 3a          10.0       1080 x 2220 dpi 420  genymotion
70eaaf75-bd9a-406d-9f01-682a5d400c6e  Samsung Galaxy S10       10.0       1440 x 3040 dpi 560  genymotion
```

Reply:
Ah, I think I see a possible complication with this approach – that's `loglevel=trace`, probably. 🤔
Nevertheless, that still might help when you get an empty list of recipes – just needs a second rerun without `--name <NAME>` filter, and the user gets help.

Reply:
See here: https://github.com/wix/Detox/pull/2575